artiq/master/worker_db.py

```diff
@@ -111,19 +111,20 @@ def __init__(self, ddb):
         self._broadcaster = Notifier(dict())
         self.local = dict()
         self.archive = dict()
+        self.metadata = dict()
 
         self.ddb = ddb
         self._broadcaster.publish = ddb.update
 
-    def set(self, key, value, broadcast=False, persist=False, archive=True):
+    def set(self, key, value, metadata={}, broadcast=False, persist=False, archive=True):
```

Comment:
When adding a new argument to a parameter list with default values, it should be added to the end (or the arguments made keyword-only) to avoid backwards-compatibility issues.

I'd also make this default to `None` and explicitly implement the default value in an `if metadata is None` check, as I've bitten one too many times by the fact that default value instances are shared between all callers across the program lifetime in Python (e.g. here all `self.metadata[key]` entries would refer to the same dict, and somebody could later fetch and change one of them, meaning to only modify that one entry).

Reply:
People might have ARTIQ forks/plugins/… – you are right that we don't care about staying backwards-compatible here, but it still good engineering practice to ensure any callsites that need to be updated fail loudly (e.g. by removing the default values as you suggested, or making the arguments keyword-only).

Reply:
removed default values